mixer/adapter/prometheus/prometheus.go

```diff
@@ -209,29 +241,77 @@ func (h *handler) HandleMetric(_ context.Context, vals []*metric.Instance) error
 				result = multierror.Append(result, fmt.Errorf("could not get value for metric %s: %v", val.Name, err))
 				continue
 			}
-			vec.With(promLabels(val.Dimensions)).Set(amt)
+			pl := promLabels(val.Dimensions)
+			if h.labelsCache != nil {
+				h.labelsCache.Set(key(val.Name, "gauge", pl), &cacheEntry{vec, pl})
+			}
+			vec.With(pl).Set(amt)
 		case config.COUNTER:
 			vec := collector.(*prometheus.CounterVec)
 			amt, err := promValue(val.Value)
 			if err != nil {
 				result = multierror.Append(result, fmt.Errorf("could not get value for metric %s: %v", val.Name, err))
 				continue
 			}
-			vec.With(promLabels(val.Dimensions)).Add(amt)
+			pl := promLabels(val.Dimensions)
+			if h.labelsCache != nil {
+				h.labelsCache.Set(key(val.Name, "counter", pl), &cacheEntry{vec, pl})
+			}
+			vec.With(pl).Add(amt)
 		case config.DISTRIBUTION:
 			vec := collector.(*prometheus.HistogramVec)
 			amt, err := promValue(val.Value)
 			if err != nil {
 				result = multierror.Append(result, fmt.Errorf("could not get value for metric %s: %v", val.Name, err))
 				continue
 			}
-			vec.With(promLabels(val.Dimensions)).Observe(amt)
+			pl := promLabels(val.Dimensions)
+			if h.labelsCache != nil {
+				h.labelsCache.Set(key(val.Name, "distribution", pl), &cacheEntry{vec, pl})
+			}
+			vec.With(pl).Observe(amt)
 		}
 	}
 
 	return result.ErrorOrNil()
 }
 
+func key(name, kind string, labels prometheus.Labels) uint64 {
+	keys := make([]string, 0, len(labels))
+	// ensure stable order
+	for k := range labels {
+		keys = append(keys, k)
+	}
+	sort.Strings(keys)
+	buf := pool.GetBuffer()
+	buf.Reset()
+	buf.WriteString(name + ":" + kind)
+	for _, k := range keys {
+		buf.WriteString(k)         // nolint: gas
+		buf.WriteString("=")       // nolint: gas
+		buf.WriteString(labels[k]) // nolint: gas
+		buf.WriteString(";")       // nolint: gas
+	}
+
+	h := fnv.New64()
+	h.Write(buf.Bytes())
```

Comment:
It is possible to flip this around? buf.Write(h)? I don't know if things can compose in this way, but it would end up being more efficient overall if possible.

Reply:
I don't see a way to do that. I improved the method a bit, refactored out the sorting, and added a benchmark test. I get the following results:

```
Running tool: /usr/local/go/bin/go test -benchmem -run=^$ istio.io/istio/mixer/adapter/prometheus -bench ^BenchmarkKey$

goos: darwin
goarch: amd64
pkg: istio.io/istio/mixer/adapter/prometheus
BenchmarkKey-8   	 1000000	      1254 ns/op	       8 B/op	       1 allocs/op
PASS
ok  	istio.io/istio/mixer/adapter/prometheus	1.417s
```

That seems fairly efficient.

Thoughts?